// async/during.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = during;

var _noop = require('lodash/noop');

var _noop2 = _interopRequireDefault(_noop);

var _onlyOnce = require('./internal/onlyOnce');

var _onlyOnce2 = _interopRequireDefault(_onlyOnce);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Like [`whilst`]{@link module:ControlFlow.whilst}, except the `test` is an asynchronous function that
 * is passed a callback in the form of `function (err, truth)`. If error is
 * passed to `test` or `fn`, the main callback is immediately called with the
 * value of the error.
 * @static
 * @method
 * execution of `fn`. Invoked with (callback).
 * The function is passed a `callback(err)`, which must be called once it has
 * completed with an optional `err` argument. Invoked with (callback).
 * function has failed and repeated execution of `fn` has stopped. `callback`
 * will be passed an error, if one occured, otherwise `null`.
 * @example
 * var count = 0;
 * async.during(
 *     function (callback) {
 *         return callback(null, count < 5);
 *     },
 *     function (callback) {
 *         count++;
 *         setTimeout(callback, 1000);
 *     },
 *     function (err) {
 *         // 5 seconds have passed
 *     }
 * );
 * @name during
 * @memberOf module:ControlFlow
 * @see [async.whilst]{@link module:ControlFlow.whilst}
 * @category Control Flow
 * @method during
 * @param {Function} test - asynchronous truth test to perform before each
 * @param {Function} fn - A function which is called each time `test` passes.
 * @param {} callback
 * @return 
 */
function during(test, fn, callback) {
    callback = (0, _onlyOnce2.default)(callback || _noop2.default);

    /**
     * Description
     * @method next
     * @param {} err
     * @return 
     */
    function next(err) {
        if (err) return callback(err);
        test(check);
    }

    /**
     * Description
     * @method check
     * @param {} err
     * @param {} truth
     * @return 
     */
    function check(err, truth) {
        if (err) return callback(err);
        if (!truth) return callback(null);
        fn(next);
    }

    test(check);
}
module.exports = exports['default'];
