// helmet-csp/lib/get-header-keys-for-browser.js
var ALL_HEADERS = require('./all-headers')

/**
 * Description
 * @method goodBrowser
 * @return ArrayExpression
 */
function goodBrowser () {
  return ['Content-Security-Policy']
}

var handlers = {
  /**
   * Description
   * @param {} browser
   * @param {} options
   * @return 
   */
  'Android Browser': function (browser, options) {
    if (parseFloat(browser.os.version) < 4.4 || options.disableAndroid) {
      return []
    } else {
      return ['Content-Security-Policy']
    }
  },

  /**
   * Description
   * @method Chrome
   * @param {} browser
   * @return 
   */
  Chrome: function (browser) {
    var version = parseFloat(browser.version)

    if (version >= 14 && version < 25) {
      return ['X-WebKit-CSP']
    } else if (version >= 25) {
      return ['Content-Security-Policy']
    } else {
      return []
    }
  },

  /**
   * Description
   * @param {} browser
   * @return 
   */
  'Chrome Mobile': function (browser) {
    if (browser.os.family === 'iOS') {
      return ['Content-Security-Policy']
    } else {
      return handlers['Android Browser'].apply(this, arguments)
    }
  },

  /**
   * Description
   * @method Firefox
   * @param {} browser
   * @return 
   */
  Firefox: function (browser) {
    var version = parseFloat(browser.version)

    if (version >= 23) {
      return ['Content-Security-Policy']
    } else if (version >= 4 && version < 23) {
      return ['X-Content-Security-Policy']
    } else {
      return []
    }
  },

  /**
   * Description
   * @method IE
   * @param {} browser
   * @return ArrayExpression
   */
  IE: function (browser) {
    var version = parseFloat(browser.version)
    var header = version < 12 ? 'X-Content-Security-Policy' : 'Content-Security-Policy'

    return [header]
  },

  'Microsoft Edge': goodBrowser,

  'Microsoft Edge Mobile': goodBrowser,

  /**
   * Description
   * @method Opera
   * @param {} browser
   * @return 
   */
  Opera: function (browser) {
    if (parseFloat(browser.version) >= 15) {
      return ['Content-Security-Policy']
    } else {
      return []
    }
  },

  /**
   * Description
   * @method Safari
   * @param {} browser
   * @return 
   */
  Safari: function (browser) {
    var version = parseFloat(browser.version)

    if (version >= 7) {
      return ['Content-Security-Policy']
    } else if (version >= 6) {
      return ['X-WebKit-CSP']
    } else {
      return []
    }
  }
}

handlers['IE Mobile'] = handlers.IE

/**
 * Description
 * @method exports
 * @param {} browser
 * @param {} options
 * @return 
 */
module.exports = function getHeaderKeysForBrowser (browser, options) {
  var handler = handlers[browser.name]

  if (handler) {
    return handler(browser, options)
  } else {
    return ALL_HEADERS
  }
}
