// boom/lib/index.js
// Load modules

var Http = require('http');
var Hoek = require('hoek');


// Declare internals

var internals = {};

/**
 * Description
 * @method wrap
 * @param {} error
 * @param {} statusCode
 * @param {} message
 * @return ConditionalExpression
 */
exports.wrap = function (error, statusCode, message) {

    Hoek.assert(error instanceof Error, 'Cannot wrap non-Error object');
    return (error.isBoom ? error : internals.initialize(error, statusCode || 500, message));
};


/**
 * Description
 * @method create
 * @param {} statusCode
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.create = function (statusCode, message, data) {

    return internals.create(statusCode, message, data, exports.create);
};

/**
 * Description
 * @method create
 * @param {} statusCode
 * @param {} message
 * @param {} data
 * @param {} ctor
 * @return error
 */
internals.create = function (statusCode, message, data, ctor) {

    var error = new Error(message ? message : undefined);       // Avoids settings null message
    Error.captureStackTrace(error, ctor);                       // Filter the stack to our external API
    error.data = data || null;
    internals.initialize(error, statusCode);
    return error;
};

/**
 * Description
 * @method initialize
 * @param {} error
 * @param {} statusCode
 * @param {} message
 * @return error
 */
internals.initialize = function (error, statusCode, message) {

    var numberCode = parseInt(statusCode, 10);
    Hoek.assert(!isNaN(numberCode) && numberCode >= 400, 'First argument must be a number (400+):', statusCode);

    error.isBoom = true;
    error.isServer = numberCode >= 500;

    if (!error.hasOwnProperty('data')) {
        error.data = null;
    }

    error.output = {
        statusCode: numberCode,
        payload: {},
        headers: {}
    };

    error.reformat = internals.reformat;
    error.reformat();

    if (!message &&
        !error.message) {

        message = error.output.payload.error;
    }

    if (message) {
        error.message = (message + (error.message ? ': ' + error.message : ''));
    }

    return error;
};


/**
 * Description
 * @method reformat
 * @return 
 */
internals.reformat = function () {

    this.output.payload.statusCode = this.output.statusCode;
    this.output.payload.error = Http.STATUS_CODES[this.output.statusCode] || 'Unknown';

    if (this.output.statusCode === 500) {
        this.output.payload.message = 'An internal server error occurred';              // Hide actual error from user
    }
    else if (this.message) {
        this.output.payload.message = this.message;
    }
};


// 4xx Client Errors

/**
 * Description
 * @method badRequest
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.badRequest = function (message, data) {

    return internals.create(400, message, data, exports.badRequest);
};


/**
 * Description
 * @method unauthorized
 * @param {} message
 * @param {} scheme
 * @param {} attributes
 * @return err
 */
exports.unauthorized = function (message, scheme, attributes) {          // Or function (message, wwwAuthenticate[])

    var err = internals.create(401, message, undefined, exports.unauthorized);

    if (!scheme) {
        return err;
    }

    var wwwAuthenticate = '';
    var i = 0;
    var il = 0;

    if (typeof scheme === 'string') {

        // function (message, scheme, attributes)

        wwwAuthenticate = scheme;

        if (attributes || message) {
            err.output.payload.attributes = {};
        }

        if (attributes) {
            var names = Object.keys(attributes);
            for (i = 0, il = names.length; i < il; ++i) {
                var name = names[i];
                if (i) {
                    wwwAuthenticate += ',';
                }

                var value = attributes[name];
                if (value === null ||
                    value === undefined) {              // Value can be zero

                    value = '';
                }
                wwwAuthenticate += ' ' + name + '="' + Hoek.escapeHeaderAttribute(value.toString()) + '"';
                err.output.payload.attributes[name] = value;
            }
        }

        if (message) {
            if (attributes) {
                wwwAuthenticate += ',';
            }
            wwwAuthenticate += ' error="' + Hoek.escapeHeaderAttribute(message) + '"';
            err.output.payload.attributes.error = message;
        }
        else {
            err.isMissing = true;
        }
    }
    else {

        // function (message, wwwAuthenticate[])

        var wwwArray = scheme;
        for (i = 0, il = wwwArray.length; i < il; ++i) {
            if (i) {
                wwwAuthenticate += ', ';
            }

            wwwAuthenticate += wwwArray[i];
        }
    }

    err.output.headers['WWW-Authenticate'] = wwwAuthenticate;

    return err;
};


/**
 * Description
 * @method forbidden
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.forbidden = function (message, data) {

    return internals.create(403, message, data, exports.forbidden);
};


/**
 * Description
 * @method notFound
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.notFound = function (message, data) {

    return internals.create(404, message, data, exports.notFound);
};


/**
 * Description
 * @method methodNotAllowed
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.methodNotAllowed = function (message, data) {

    return internals.create(405, message, data, exports.methodNotAllowed);
};


/**
 * Description
 * @method notAcceptable
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.notAcceptable = function (message, data) {

    return internals.create(406, message, data, exports.notAcceptable);
};


/**
 * Description
 * @method proxyAuthRequired
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.proxyAuthRequired = function (message, data) {

    return internals.create(407, message, data, exports.proxyAuthRequired);
};


/**
 * Description
 * @method clientTimeout
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.clientTimeout = function (message, data) {

    return internals.create(408, message, data, exports.clientTimeout);
};


/**
 * Description
 * @method conflict
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.conflict = function (message, data) {

    return internals.create(409, message, data, exports.conflict);
};


/**
 * Description
 * @method resourceGone
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.resourceGone = function (message, data) {

    return internals.create(410, message, data, exports.resourceGone);
};


/**
 * Description
 * @method lengthRequired
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.lengthRequired = function (message, data) {

    return internals.create(411, message, data, exports.lengthRequired);
};


/**
 * Description
 * @method preconditionFailed
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.preconditionFailed = function (message, data) {

    return internals.create(412, message, data, exports.preconditionFailed);
};


/**
 * Description
 * @method entityTooLarge
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.entityTooLarge = function (message, data) {

    return internals.create(413, message, data, exports.entityTooLarge);
};


/**
 * Description
 * @method uriTooLong
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.uriTooLong = function (message, data) {

    return internals.create(414, message, data, exports.uriTooLong);
};


/**
 * Description
 * @method unsupportedMediaType
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.unsupportedMediaType = function (message, data) {

    return internals.create(415, message, data, exports.unsupportedMediaType);
};


/**
 * Description
 * @method rangeNotSatisfiable
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.rangeNotSatisfiable = function (message, data) {

    return internals.create(416, message, data, exports.rangeNotSatisfiable);
};


/**
 * Description
 * @method expectationFailed
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.expectationFailed = function (message, data) {

    return internals.create(417, message, data, exports.expectationFailed);
};

/**
 * Description
 * @method badData
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.badData = function (message, data) {

    return internals.create(422, message, data, exports.badData);
};


/**
 * Description
 * @method preconditionRequired
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.preconditionRequired = function (message, data) {

    return internals.create(428, message, data, exports.preconditionRequired);
};


/**
 * Description
 * @method tooManyRequests
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.tooManyRequests = function (message, data) {

    return internals.create(429, message, data, exports.tooManyRequests);
};


// 5xx Server Errors

/**
 * Description
 * @method internal
 * @param {} message
 * @param {} data
 * @param {} statusCode
 * @return CallExpression
 */
exports.internal = function (message, data, statusCode) {

    return internals.serverError(message, data, statusCode, exports.internal);
};

/**
 * Description
 * @method serverError
 * @param {} message
 * @param {} data
 * @param {} statusCode
 * @param {} ctor
 * @return error
 */
internals.serverError = function (message, data, statusCode, ctor) {

    var error;
    if (data instanceof Error) {
        error = exports.wrap(data, statusCode, message);
    } else {
        error = internals.create(statusCode || 500, message, undefined, ctor);
        error.data = data;
    }

    return error;
};


/**
 * Description
 * @method notImplemented
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.notImplemented = function (message, data) {

    return internals.serverError(message, data, 501, exports.notImplemented);
};


/**
 * Description
 * @method badGateway
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.badGateway = function (message, data) {

    return internals.serverError(message, data, 502, exports.badGateway);
};


/**
 * Description
 * @method serverTimeout
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.serverTimeout = function (message, data) {

    return internals.serverError(message, data, 503, exports.serverTimeout);
};


/**
 * Description
 * @method gatewayTimeout
 * @param {} message
 * @param {} data
 * @return CallExpression
 */
exports.gatewayTimeout = function (message, data) {

    return internals.serverError(message, data, 504, exports.gatewayTimeout);
};


/**
 * Description
 * @method badImplementation
 * @param {} message
 * @param {} data
 * @return err
 */
exports.badImplementation = function (message, data) {

    var err = internals.serverError(message, data, 500, exports.badImplementation);
    err.isDeveloperError = true;
    return err;
};
