// core-util-is/lib/util.js
// Copyright Joyent, Inc. and other Node contributors.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.

// NOTE: These type checking functions intentionally don't use `instanceof`
// because it is fragile and can be easily faked with `Object.create()`.

/**
 * Description
 * @method isArray
 * @param {} arg
 * @return BinaryExpression
 */
function isArray(arg) {
  if (Array.isArray) {
    return Array.isArray(arg);
  }
  return objectToString(arg) === '[object Array]';
}
exports.isArray = isArray;

/**
 * Description
 * @method isBoolean
 * @param {} arg
 * @return BinaryExpression
 */
function isBoolean(arg) {
  return typeof arg === 'boolean';
}
exports.isBoolean = isBoolean;

/**
 * Description
 * @method isNull
 * @param {} arg
 * @return BinaryExpression
 */
function isNull(arg) {
  return arg === null;
}
exports.isNull = isNull;

/**
 * Description
 * @method isNullOrUndefined
 * @param {} arg
 * @return BinaryExpression
 */
function isNullOrUndefined(arg) {
  return arg == null;
}
exports.isNullOrUndefined = isNullOrUndefined;

/**
 * Description
 * @method isNumber
 * @param {} arg
 * @return BinaryExpression
 */
function isNumber(arg) {
  return typeof arg === 'number';
}
exports.isNumber = isNumber;

/**
 * Description
 * @method isString
 * @param {} arg
 * @return BinaryExpression
 */
function isString(arg) {
  return typeof arg === 'string';
}
exports.isString = isString;

/**
 * Description
 * @method isSymbol
 * @param {} arg
 * @return BinaryExpression
 */
function isSymbol(arg) {
  return typeof arg === 'symbol';
}
exports.isSymbol = isSymbol;

/**
 * Description
 * @method isUndefined
 * @param {} arg
 * @return BinaryExpression
 */
function isUndefined(arg) {
  return arg === void 0;
}
exports.isUndefined = isUndefined;

/**
 * Description
 * @method isRegExp
 * @param {} re
 * @return BinaryExpression
 */
function isRegExp(re) {
  return objectToString(re) === '[object RegExp]';
}
exports.isRegExp = isRegExp;

/**
 * Description
 * @method isObject
 * @param {} arg
 * @return LogicalExpression
 */
function isObject(arg) {
  return typeof arg === 'object' && arg !== null;
}
exports.isObject = isObject;

/**
 * Description
 * @method isDate
 * @param {} d
 * @return BinaryExpression
 */
function isDate(d) {
  return objectToString(d) === '[object Date]';
}
exports.isDate = isDate;

/**
 * Description
 * @method isError
 * @param {} e
 * @return LogicalExpression
 */
function isError(e) {
  return (objectToString(e) === '[object Error]' || e instanceof Error);
}
exports.isError = isError;

/**
 * Description
 * @method isFunction
 * @param {} arg
 * @return BinaryExpression
 */
function isFunction(arg) {
  return typeof arg === 'function';
}
exports.isFunction = isFunction;

/**
 * Description
 * @method isPrimitive
 * @param {} arg
 * @return LogicalExpression
 */
function isPrimitive(arg) {
  return arg === null ||
         typeof arg === 'boolean' ||
         typeof arg === 'number' ||
         typeof arg === 'string' ||
         typeof arg === 'symbol' ||  // ES6 symbol
         typeof arg === 'undefined';
}
exports.isPrimitive = isPrimitive;

exports.isBuffer = Buffer.isBuffer;

/**
 * Description
 * @method objectToString
 * @param {} o
 * @return CallExpression
 */
function objectToString(o) {
  return Object.prototype.toString.call(o);
}
