// caseless/index.js
/**
 * Description
 * @method Caseless
 * @param {} dict
 * @return 
 */
function Caseless (dict) {
  this.dict = dict || {}
}
/**
 * Description
 * @method set
 * @param {} name
 * @param {} value
 * @param {} clobber
 * @return 
 */
Caseless.prototype.set = function (name, value, clobber) {
  if (typeof name === 'object') {
    for (var i in name) {
      this.set(i, name[i], value)
    }
  } else {
    if (typeof clobber === 'undefined') clobber = true
    var has = this.has(name)

    if (!clobber && has) this.dict[has] = this.dict[has] + ',' + value
    else this.dict[has || name] = value
    return has
  }
}
/**
 * Description
 * @method has
 * @param {} name
 * @return Literal
 */
Caseless.prototype.has = function (name) {
  var keys = Object.keys(this.dict)
    , name = name.toLowerCase()
    ;
  for (var i=0;i<keys.length;i++) {
    if (keys[i].toLowerCase() === name) return keys[i]
  }
  return false
}
/**
 * Description
 * @method get
 * @param {} name
 * @return result
 */
Caseless.prototype.get = function (name) {
  name = name.toLowerCase()
  var result, _key
  var headers = this.dict
  Object.keys(headers).forEach(function (key) {
    _key = key.toLowerCase()
    if (name === _key) result = headers[key]
  })
  return result
}
/**
 * Description
 * @method swap
 * @param {} name
 * @return 
 */
Caseless.prototype.swap = function (name) {
  var has = this.has(name)
  if (!has) throw new Error('There is no header than matches "'+name+'"')
  this.dict[name] = this.dict[has]
  delete this.dict[has]
}
/**
 * Description
 * @method del
 * @param {} name
 * @return UnaryExpression
 */
Caseless.prototype.del = function (name) {
  var has = this.has(name)
  return delete this.dict[has || name]
}

/**
 * Description
 * @method exports
 * @param {} dict
 * @return NewExpression
 */
module.exports = function (dict) {return new Caseless(dict)}
/**
 * Description
 * @method httpify
 * @param {} resp
 * @param {} headers
 * @return c
 */
module.exports.httpify = function (resp, headers) {
  var c = new Caseless(headers)
  /**
   * Description
   * @method setHeader
   * @param {} key
   * @param {} value
   * @param {} clobber
   * @return CallExpression
   */
  resp.setHeader = function (key, value, clobber) {
    if (typeof value === 'undefined') return
    return c.set(key, value, clobber)
  }
  /**
   * Description
   * @method hasHeader
   * @param {} key
   * @return CallExpression
   */
  resp.hasHeader = function (key) {
    return c.has(key)
  }
  /**
   * Description
   * @method getHeader
   * @param {} key
   * @return CallExpression
   */
  resp.getHeader = function (key) {
    return c.get(key)
  }
  /**
   * Description
   * @method removeHeader
   * @param {} key
   * @return CallExpression
   */
  resp.removeHeader = function (key) {
    return c.del(key)
  }
  resp.headers = c.dict
  return c
}
