// async/internal/DoublyLinkedList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = DLL;
// Simple doubly linked list (https://en.wikipedia.org/wiki/Doubly_linked_list) implementation
// used for queues. This implementation assumes that the node provided by the user can be modified
// to adjust the next and last properties. We implement only the minimal functionality
// for queue support.
/**
 * Description
 * @method DLL
 * @return 
 */
function DLL() {
    this.head = this.tail = null;
    this.length = 0;
}

/**
 * Description
 * @method setInitial
 * @param {} dll
 * @param {} node
 * @return 
 */
function setInitial(dll, node) {
    dll.length = 1;
    dll.head = dll.tail = node;
}

/**
 * Description
 * @method removeLink
 * @param {} node
 * @return node
 */
DLL.prototype.removeLink = function (node) {
    if (node.prev) node.prev.next = node.next;else this.head = node.next;
    if (node.next) node.next.prev = node.prev;else this.tail = node.prev;

    node.prev = node.next = null;
    this.length -= 1;
    return node;
};

DLL.prototype.empty = DLL;

/**
 * Description
 * @method insertAfter
 * @param {} node
 * @param {} newNode
 * @return 
 */
DLL.prototype.insertAfter = function (node, newNode) {
    newNode.prev = node;
    newNode.next = node.next;
    if (node.next) node.next.prev = newNode;else this.tail = newNode;
    node.next = newNode;
    this.length += 1;
};

/**
 * Description
 * @method insertBefore
 * @param {} node
 * @param {} newNode
 * @return 
 */
DLL.prototype.insertBefore = function (node, newNode) {
    newNode.prev = node.prev;
    newNode.next = node;
    if (node.prev) node.prev.next = newNode;else this.head = newNode;
    node.prev = newNode;
    this.length += 1;
};

/**
 * Description
 * @method unshift
 * @param {} node
 * @return 
 */
DLL.prototype.unshift = function (node) {
    if (this.head) this.insertBefore(this.head, node);else setInitial(this, node);
};

/**
 * Description
 * @method push
 * @param {} node
 * @return 
 */
DLL.prototype.push = function (node) {
    if (this.tail) this.insertAfter(this.tail, node);else setInitial(this, node);
};

/**
 * Description
 * @method shift
 * @return LogicalExpression
 */
DLL.prototype.shift = function () {
    return this.head && this.removeLink(this.head);
};

/**
 * Description
 * @method pop
 * @return LogicalExpression
 */
DLL.prototype.pop = function () {
    return this.tail && this.removeLink(this.tail);
};
module.exports = exports['default'];
